internal/extensions/extensions.go

```diff
@@ -0,0 +1,213 @@
+		}
+	}
+
+	return &extensions, nil
+}
+
+// NewExtensionRegistryFromList creates a new instance of the Extensions struct, from a list of extensions.
+// The passed extensions could be internal or external or just invalid.
+// The function is typically used to create a registry from an API response in order to compare
+// the source and the target systems' extensions for compatibility.
+// The returned registry is always sorted with internal extensions first.
+func NewExtensionRegistryFromList(extensions []string) (*Extensions, error) {
+	registry, err := NewExtensionRegistry(false)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, extension := range extensions {
+		err = validateExternalExtension(extension)
+		if err != nil {
+			// The extension could be an internal extension.
+			err = validateInternalExtension(extension)
+			if err != nil {
+				// The extension is invalid in both case.
+				return nil, fmt.Errorf("Extension %q is invalid", extension)
+			}
+
+			// The extension is an internal extension.
+			*registry = append(*registry, extension)
+		} else {
+			// The extension is an external extension.
+			*registry = append(*registry, extension)
+		}
+	}
+
+	return registry, nil
+}
+
+// Register registers new external extensions to the Extensions struct.
+func (e *Extensions) Register(newExtensions []string) error {
+	// Check for duplicates for internal and External extensions
+	for _, extension := range newExtensions {
+		if shared.ValueInSlice[string](extension, []string(*e)) {
+			return fmt.Errorf("Extension %q already registered", extension)
+		}
+
+		err := validateExternalExtension(extension)
+		if err != nil {
+			return err
+		}
+
+		*e = append(*e, extension)
+	}
+
+	return nil
+}
+
+// MarshalJSON implement the json.Marshaler interface to serialize the Extensions struct
+// for API responses.
+func (e *Extensions) MarshalJSON() ([]byte, error) {
+	return json.Marshal([]string(*e))
+}
+
+// UnmarshalJSON implement the json.Unmarshaler interface to deserialize the Extensions struct
+// from API requests.
+func (e *Extensions) UnmarshalJSON(data []byte) error {
+	var tmp []string
+	if err := json.Unmarshal(data, &tmp); err != nil {
+		return err
+	}
+
+	*e = Extensions(tmp)
+	return nil
+}
+
+// Value implements the driver.Valuer interface to serialize the Extensions struct for database storage.
+func (e *Extensions) Value() (driver.Value, error) {
+	if e == nil {
+		return nil, nil
+	}
+
+	return json.Marshal(e)
+}
+
+// Scan implements the sql.Scanner interface to deserialize the Extensions struct from database storage.
+func (e *Extensions) Scan(value interface{}) error {
+	if value == nil {
+		*e = nil
+		return nil
+	}
+
+	bytes, ok := value.([]byte)
+	if !ok {
+		return errors.New("type assertion to []byte failed, incompatible type")
+	}
+
+	return json.Unmarshal(bytes, e)
+}
+
+// IsSameVersion checks if the source registry supports the target registry.
+func (e *Extensions) IsSameVersion(t *Extensions) error {
```

Comment:
Since `Extensions` is a slice, it is already a pointer. I think most of these methods can have the value rather the reference as the receiver.

The `UnmarshalJSON` implementation would be an exception as the `json` library must be given a reference, but that method may or may not be needed (see above).

Reply:
fixed